### PR TITLE
Fixed potential UB in edge cache migration

### PR DESCRIPTION
**Pull request by jsteemann**
### Scope & Purpose

* Fixed potential undefined behavior in edge cache during cache migration tasks.
  There was a short window of time in which an already freed Table could be
  used by concurrently running edge lookups.

The bug can happen when using the edge cache and at the same time a migration task for the same cache is executed.
This requires that concurrent cache access(es) and a cache migration happen for the same edge collection/shard's edge cache at the same time. And still then it is a matter of luck/misfortune because there is only a small window of time in which a raw pointer to the cache's Table is used and dereferenced, while the migration task deletes this very same Table.
A migration to a larger cache is kicked off when after a cache insert the cache goes above 25% fill grade. A migration is also kicked off on a remove if the cache goes to a fill grade less than 4%.

- [x] :hankey: Bugfix (requires CHANGELOG entry)
- [ ] :pizza: New feature (requires CHANGELOG entry, feature documentation and release notes)
- [ ] :fire: Performance improvement
- [ ] :hammer: Refactoring/simplification
- [x] :book: CHANGELOG entry made

#### Backports:

- [x] Backport for 3.7: https://github.com/arangodb/arangodb/pull/15050

### Testing & Verification

- [x] The behavior in this PR was *manually tested*
- [x] This change is already covered by existing tests, such as *gtest*.

Additionally tested via setting a local cluster with the edge cache limited to 4 MB (`--cache.size 4000000`) and running the following script repeatedly:
```bash 
scripts/startLocalCluster.sh -d 3 -a 1 # starts cluster on port 8530
colls=3
for i in `seq 1 $colls`; do 
  echo "cn = \"testi$i\"; try { c = db._createEdgeCollection(cn, { numberOfShards: 40, replicationFactor: 1 }); } catch (err) { c = db._collection(cn); } docs = []; for (i = 0; i < 1000000; ++i) { docs.push({ _from: \"v1/testi\" + Math.floor(Math.random() * 100000), _to: \"v2/testi\" + Math.floor(Math.random() * 1000000) }); if (docs.length === 5000) { print(i); c.insert(docs); docs = []; } }" | build/bin/arangosh --server.endpoint tcp://127.0.0.1:8530 & 
done
sleep 4
for i in `seq 1 $(($colls * 2))`; do 
  echo "cn = \"testi\" + (($i % $colls) + 1); c = db._collection(cn); while (true) { docs = []; for (i = 0; i < 10000; ++i) { docs.push(\"v1/testi\" + Math.floor(Math.random() * 100000)); } db._query('FOR doc IN @@cn FILTER doc._from IN @docs RETURN doc', { '@cn': cn, docs }, { silent: true }); }" | build/bin/arangosh --server.endpoint tcp://127.0.0.1:8530 & 
done
for i in `seq 1 $(($colls * 2))`; do 
  echo "cn = \"testi\" + (($i % $colls) + 1); c = db._collection(cn); while (true) { docs = []; for (i = 0; i < 10000; ++i) { docs.push(\"v2/testi\" + Math.floor(Math.random() * 1000000)); } db._query('FOR doc IN @@cn FILTER doc._to IN @docs RETURN doc', { '@cn': cn, docs }, { silent: true }); }" | build/bin/arangosh --server.endpoint tcp://127.0.0.1:8530 & 
done
```

### CODE REVIEW

**Review by mpoeter — COMMENTED** [review]
One comment, otherwise LGTM.

Do we have a testcase that triggers the issue that we can add to the (chaos) integration tests?

File: arangod/Cache/Cache.cpp [hunk 1]
```diff
@@ -337,7 +337,8 @@ bool Cache::reportInsert(bool hadEviction) {
 Metadata& Cache::metadata() { return _metadata; }
 
 std::shared_ptr<Table> Cache::table() const {
-  return std::atomic_load(&_tableShrdPtr);
+  // TODO: check if std::atomic_load_explicitt(&_table, std::memory_order_relaxed); works here!
+  return std::atomic_load(&_table);
```
[the remark this answers is not in the document]
Comment (jsteemann) [reply arangod/Cache/Cache.cpp:341]: Ok, I can try it.

[the remark this answers is not in the document]
Comment (jsteemann) [reply arangod/Cache/Cache.cpp]: Done in this PR and the PRs for the other branches.

[the remark this answers is not in the document]
Comment (mpoeter) [reply arangod/Cache/Cache.cpp]: Good! In addition we should also adjust the memory order of the store operations to use `memory_order_release`.

**Review by mpoeter — APPROVED** [review]

File: arangod/Cache/Cache.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (jsteemann) [reply arangod/Cache/Cache.cpp]: Ok, pushed here and in the other branches.

**Pull request merged**